Automattic/jetpack
# Sitemaps: remove post limit on sitemap

Addresses #3314: Remove 1000 post limit on sitemaps.
Addresses #3750: Support video sitemaps (in progress)

#### Changes proposed in this Pull Request:
Replace sitemap generating code.

Note that this PR is an update of #5394, which became broken because I am awesome at git

Documentation is on the Team Luna P2: p7rd6c-B0-p2
Also some design ideas: p7rd6c-fS-p2

#### Testing instructions:
1. Create some posts (see [FakerPress](https://wordpress.org/plugins/fakerpress/)). Make sure some of them have a recent (<=2 day old) timestamp.
2. Activate the sitemaps module of Jetpack.
3. Set permalinks to "pretty" (doesn't matter what structure)
4. Go to `example.com/news-sitemap.xml`.
5. Go to `example.com/sitemap.xml`. (It may take a refresh or two the first time.)
6. Set permalinks to "plain"
7. Go to `example.com/?jetpack-sitemap=sitemap.xml` and `example.com/?jetpack-sitemap=news-sitemap.xml`
8. Add some images and videos to the media library and repeat steps 3--7.

These changes allow for arbitrarily large sitemap trees per the [spec](https://www.sitemaps.org). To test this functionality without having to create an arbitrarily large site, adjust the constant `SITEMAP_MAX_ITEMS` in `modules/sitemaps/sitemap-buffer.php`, (5 is a good value.) This constant determines the maximum number of URLs on a single sitemap, so that if it's set to 5 and we have 6 posts/pages then the tree structure kicks in.

The constant `SITEMAP_INTERVAL` in `modules/sitemaps/sitemaps.php` is the minimum number of seconds between rebuilds; it's currently set to 60 for testing.

Past related issues/PRs:
* #4093 (Support PATHINFO permalinks): This module should, at the URL `example.com/index.php/sitemap.xml`.
* #4056 (Decode non-XML entities): We still run stylesheet strings through `ent2ncr` to avoid this bug.


#### Proposed changelog entry for your changes:

New filters:
* `jetpack_sitemap_image_skip_post`: For selectively excluding images from the sitemap.
* `jetpack_sitemap_image_sitemap_item`: For manipulating the XML of image entries in the sitemap.
* `jetpack_sitemap_video_skip_post`: For selectively excluding videos from the sitemap.
* `jetpack_sitemap_video_sitemap_item`: For manipulating the XML of video entries in the sitemap.
* `jetpack_sitemap_news_ns`: Add XML namespaces for the news sitemap.
* `jetpack_sitemap_image_ns`: Add XML namespaces for image sitemaps.

Filters with changed behavior:
* `jetpack_sitemap_location`: The old version had only two sitemap URLs to keep track of, so it was feasible to make both of them configurable. This version keeps track of two URLs and 6 indexed families of unknown size. Do we make them all configurable?

---
